### ISSUE TITLE
Add --chown to WORKDIR

### ISSUE THREAD

**Issue by omercnet** [moby/moby#36408]
In a lot of Dockerfiles you see WORKDIR followed by `RUN chown 1000:1000 $(pwd)`

It only makes sense to combine the two just like was done for ADD/COPY in #34263 to allow creating the WORKDIR with appropriate ownership

**Comment by thaJeztah** [comment]
Prior issues/discussions on this: https://github.com/moby/moby/issues/20295 (https://github.com/moby/moby/issues/20295#issuecomment-254951672). 

Also, to be sure: checking history on the WORKDIR option, it looks like this always created the directory as `root`, never taking `USER` into account(https://github.com/moby/moby/pull/1459 / https://github.com/moby/moby/commit/687d27ab575778196ea646b6d3fa61b6c8e512b0)

**Comment by omercnet** [comment]
It's not about respecting `USER` (which would be convenient, actually), it's about adding a `--chown` flag like that was recently introduced for ADD/COPY

**Comment by liuchong** [comment]
> Discussing this in the maintainers meeting, and the difference with `COPY` and `ADD` is that for `WORKDIR` you have the option to first create the directory, before setting it as workdir (which is generally the recommendation), so we don't think there's a need to add this as a feature, something like:
> 
> ```dockerfile
> USER foo
> RUN mkdir /my/workdir
> WORKDIR /my/workdir
> ```
> 
> For that reason, I'm closing this for now, but feel free to discuss

I don't think this is very helpful, as command `RUN` run as user `foo`, it may has no permission of directory `/my`, so you still need create that directory first as root or chown. Seems the only safe place without permission problem is `/home/foo`.

But still if we use a USER foo, we need create the user(/group) first, the process is so lengthy 😹

**Comment by omercnet** [comment]
I appreciate the need to create clear Dockerfiles with enough verbosity so readers can understand what's being done - same reason you'll see most maintainers use `--long-params` instead of short `-l` but we need to make sure this principal doesn't create lengthy Dockerfiles that are just a hassle to maintain

with respect to your response:
> Discussing this in the maintainers meeting, and the difference with `COPY` and `ADD` is that for `WORKDIR` you have the option to first create the directory, before setting it as workdir (which is generally the recommendation), so we don't think there's a need to add this as a feature, something like:
> 
> ```dockerfile
> USER foo
> RUN mkdir /my/workdir
> WORKDIR /my/workdir
> ```
> 
> For that reason, I'm closing this for now, but feel free to discuss

you can claim that someone could `mkdir` before `COPY` or `ADD` as well, resulting in
```
RUN mkdir /src
COPY . /src
RUN chown foo:foo /src
USER foo
```
but it makes much more sense replacing these 4 instructions with `COPY --chown foo:foo . /src` and so does `WORKDIR` deserve a `--chown` as well to keep things simple and comparable with other instructions

**Comment by thaJeztah** [comment]
> but it makes much more sense replacing these 4 instructions with `COPY --chown foo:foo . /src`

There's a difference in that `COPY` takes files from the build-context and changes their ownership (as ownership is reset before the `COPY` is performed, and therefore cannot be set otherwise). w.r.t. creating the `/src` directory before `COPY`; this would actually be recommended, given that (depending on if `COPY` is copying file(s) or a directory), `/src`  will be created as `root` before the copy takes place (more discussion about that in https://github.com/moby/moby/issues/35639 and https://github.com/moby/moby/issues/38710)

**Comment by APoole00** [comment]
1. Fix WORKDIR to create the directory as the current user by default. This is clearly the expected behavior from developers. 
2. Increase the rev to indicate a breaking change and document the change.

This is not nearly as big of an issue as you make it out to be. Most of us have now read this and put RUN mkdir in before WORKDIR after painfully finding out that our code wasn't working because WORKDIR makes everything as root. The only case this change would break would be the case where the user was set as something other than root. Then added a directory while being that other user using WORKDIR that they did not want the current user to have access to. That seems like such an edge case. I can't think of any reason someone would knowing do such a thing. However, I know that many of us have unknowing done just that because we are creating tickets because the behavior of WORKDIR is that unintuitive.

**Comment by nodesocket** [comment]
Currently doing this:

```docker
USER python
RUN mkdir /home/python/app
WORKDIR /home/python/app
```

which is not nearly as easy to understand and reason about then if the `--chown` flag existed:

```docker
WORKDIR --chown=python /home/python/app
USER python
```

Please add support for `--chown` with `WORKDIR`.

**Comment by thaJeztah** [comment]
@nodesocket best to open a feature request in https://github.com/moby/buildkit (which is where the Dockerfile syntax / specification is maintained).

That said; I'm not sure if `--chown` would be the best flag for this, as it may give the impression that `WORKDIR` is always `chown`'d, so this could lead to the expectation that the Dockerfile below would (recursively?) `chown` the `WORKDIR` multiple times;


```dockerfile
FROM foo
RUN mkdir /this/is/some/directory && chown -R foo:bar /this

WORKDIR --chown bar:baz /this/is/some
RUN echo foo

WORKDIR --chown user:group /this/is
RUN echo bar
```

**Comment by mthalman** [comment]
This is even more relevant with the rise in popularity of distroless containers. Simply running "mkdir" can't be done in a distroless stage since there's no shell. So you're forced to do it in a stage of a Dockerfile that does have a shell. And then you have to do a COPY to get that where you need. And then you may end up running into https://github.com/moby/moby/issues/38710. All unncessary complexity that could be solved by being able to set ownership directly from the WORKDIR instruction.

**Comment by APoole00** [comment]
I really hope you keep metrics on how many people view this issue and not just comment. There's a real demand for this ability and I think this should be opened and worked.

**Comment by thaJeztah** [comment]
> Simply running "mkdir" can't be done in a distroless stage since there's no shell. So you're forced to do it in a stage of a Dockerfile that does have a shell.

Have you tried with BuildKit enabled? I think setting `USER` before the `WORKDIR` should create  the `WORKDIR` with the current user as owner.

> I really hope you keep metrics on how many people view this issue and not just comment. There's a real demand for this ability and I think this should be opened and worked.

Thank you for commenting; please take the time to scroll up 2 comments: https://github.com/moby/moby/issues/36408#issuecomment-1193814135

> best to open a feature request in https://github.com/moby/buildkit (which is where the Dockerfile syntax / specification is maintained).

**Comment by kiranpradeep** [comment]
> Have you tried with BuildKit enabled? I think setting USER before the WORKDIR should create the WORKDIR with the current user as owner.

If we follow this, the problem moves to ADD. ADD (unlike COPY) does not have chown. If we use ADD (for urls, tar files), then we need to follow it with _RUN chown_. That will now fail since USER before WORKDIR sets to standard user who don't have rights to _chown_. Then we need to switch back to _USER root_ after ADD and then again ..

**Comment by gimmi** [comment]
> I think setting USER before the WORKDIR should create the WORKDIR with the current user as owner.

@thaJeztah Confirm this is the behavior I see happening when `WORKDIR` dont already exists, but didn't find documentation anywhere.

**Comment by thaJeztah** [comment]
@gimmi documentation for the Dockerfile syntax is maintained in the https://github.com/moby/buildkit project; if documentation is missing on this part, it's best to open a ticket there. The related file is found here; https://github.com/moby/buildkit/blob/v0.21/frontend/dockerfile/docs/reference.md#workdir
